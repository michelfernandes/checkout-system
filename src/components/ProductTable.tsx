import * as React from "react";
import CRUDTable,
{
  Fields,
  Field,
  CreateForm,
} from "react-crud-table";
import type { Product } from '../App'

interface ProductTableProps {
  products: Product[],
  setProducts: (products: Product[]) => void
}

export const ProductTable: React.FC<ProductTableProps> = ({
  products,
  setProducts
}) => {

  let count = products.length;
  const service = {
    fetchItems: () => {
      let result = Array.from(products);
      return Promise.resolve(result);
    },
    create: (product) => {
      count += 1;
      setProducts([...products,{...product,id:count}])
      return Promise.resolve(product);
    },
    delete: (data) => {
      const product = products.find(t => t.id === data.id);
      const filteredproducts = products.filter(t => t.id !== product!.id);
      setProducts(filteredproducts);
      return Promise.resolve(product);
    },
  };

return(
  <div>
    <CRUDTable
      caption="Products"
      fetchItems={() => service.fetchItems()}
    >
      <Fields>
        <Field
          name="id"
          label="Id"
          hideInCreateForm
          readOnly
        />
        <Field
          name="name"
          label="Name"
        />
        <Field
          name="type"
          label="Type"
        />
        <Field
          name="price"
          label="Price"
          type="number"
        />
      </Fields>
      <CreateForm
        title="Create a new product!"
        message=""
        trigger="Create product"
        onSubmit={product => service.create(product)}
        submitText="Create"
      />
    </CRUDTable>
  </div>
);
      }