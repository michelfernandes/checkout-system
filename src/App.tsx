import * as React from 'react';
import './App.css';
import { ProductTable } from './components/ProductTable';
import { PromotionalRulesList } from './components/PromotionalRulesList';
import { CheckoutTable, CheckoutScan } from './components/Checkout';

export type Product = {
  id: number,
  name: string,
  type: string,
  price: number
}

function App() {
  
  
  const [products, setProducts] = React.useState<Product[]>([
    {
      id: 1,
      name: "Curry Sauce",
      type: "Food",
      price: 1.95,
    },
    {
      id: 2,
      name: "Pizza",
      type: "Food",
      price: 5.99,
    },
    {
      id: 3,
      name: "Menâ€™s T-Shirt ",
      type: "Clothing",
      price: 25.00,
    }
  ]);

  const [totalPrice, setTotalPrice] = React.useState(0);
  const [scannedProducts, setScannedProducts] = React.useState<Product[]>([]);

  return (
    <>
      <div style={{textAlign: 'center'}}>
        <h2>Checkout System</h2>
      </div>
      <div style={{display: 'flex'}}>
        <div className="section-container">
          <ProductTable products={products} setProducts={setProducts}/>
        </div>
        <div className="section-container">
            <PromotionalRulesList/>
        </div>
        <div className="section-container">
          <CheckoutScan products={products} scannedProducts={scannedProducts} setScannedProducts={setScannedProducts} setTotalPrice={setTotalPrice}/>
          <CheckoutTable scannedProducts={scannedProducts} totalPrice={totalPrice} />
        </div>
      </div>
    </>
  );
}

export default App;
