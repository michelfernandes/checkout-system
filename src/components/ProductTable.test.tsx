import { render } from '@testing-library/react';
import { Product } from '../App';
import { ProductTable } from './ProductTable';

test('Should match Product Table snapshot', () => {
  const productTable = render(<ProductTable products={[]} setProducts={function (products: Product[]): void {
    throw new Error('Function not implemented.');
  } } />);
  expect(productTable).toMatchSnapshot();
});
