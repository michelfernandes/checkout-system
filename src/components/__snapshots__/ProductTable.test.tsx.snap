// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match Product Table snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="section-caption"
      >
        <span>
          Products
        </span>
      </div>
      <div>
        <div>
          <button
            class="crud-button crud-button--positive crud-button--modal-trigger"
          >
            Create product
          </button>
          <div
            class="crud-modal-wrapper"
            style="display: none;"
          >
            <div
              class="crud-modal-wrapper__background"
            />
            <div
              class="crud-modal-wrapper__modal"
            >
              <button
                class="crud-button crud-button--circular crud-button--close"
              >
                X
              </button>
              <h3
                class="crud-modal-wrapper__title"
              >
                Create a new product!
              </h3>
              <div>
                <form
                  class="crud-modal-form"
                >
                  
                  <div
                    class="crud-modal-form__field-container"
                  >
                    <label
                      class="crud-modal-form__label"
                      for="name"
                    >
                      Name
                    </label>
                    <input
                      name="name"
                      type="text"
                      value=""
                    />
                  </div>
                  <div
                    class="crud-modal-form__field-container"
                  >
                    <label
                      class="crud-modal-form__label"
                      for="type"
                    >
                      Type
                    </label>
                    <input
                      name="type"
                      type="text"
                      value=""
                    />
                  </div>
                  <div
                    class="crud-modal-form__field-container"
                  >
                    <label
                      class="crud-modal-form__label"
                      for="price"
                    >
                      Price
                    </label>
                    <input
                      name="price"
                      type="number"
                      value=""
                    />
                  </div>
                  <button
                    class="crud-button crud-button--positive"
                    type="submit"
                  >
                    Create
                  </button>
                </form>
              </div>
            </div>
          </div>
        </div>
        <div
          class="crud-table__caption"
        >
          Products
        </div>
        <table
          class="crud-table"
        >
          <thead
            class="crud-table__header"
          >
            <tr
              class="crud-table__row crud-table__row--fields"
            >
              <th
                class="crud-table__header-cell"
              >
                Id
                 
                <span>
                  ▼
                </span>
              </th>
              <th
                class="crud-table__header-cell"
              >
                Name
                 
              </th>
              <th
                class="crud-table__header-cell"
              >
                Type
                 
              </th>
              <th
                class="crud-table__header-cell"
              >
                Price
                 
              </th>
              
            </tr>
            <tr
              class="crud-table__row crud-table__row--sorter"
            >
              <th
                class="crud-table__header-cell"
                colspan="4"
              >
                Sort by: 
                <br />
                <select
                  class="crud-table__select"
                >
                  <option
                    value=""
                  >
                    Select field
                  </option>
                  <option
                    value="id"
                  >
                    Id
                  </option>
                  <option
                    value="name"
                  >
                    Name
                  </option>
                  <option
                    value="type"
                  >
                    Type
                  </option>
                  <option
                    value="price"
                  >
                    Price
                  </option>
                </select>
                 
                <select
                  class="crud-table__select"
                >
                  <option
                    value=""
                  >
                    Select direction
                  </option>
                  <option
                    value="descending"
                  >
                    descending
                  </option>
                  <option
                    value="ascending"
                  >
                    ascending
                  </option>
                </select>
              </th>
            </tr>
          </thead>
          <tbody
            class="crud-table__body"
          />
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="section-caption"
    >
      <span>
        Products
      </span>
    </div>
    <div>
      <div>
        <button
          class="crud-button crud-button--positive crud-button--modal-trigger"
        >
          Create product
        </button>
        <div
          class="crud-modal-wrapper"
          style="display: none;"
        >
          <div
            class="crud-modal-wrapper__background"
          />
          <div
            class="crud-modal-wrapper__modal"
          >
            <button
              class="crud-button crud-button--circular crud-button--close"
            >
              X
            </button>
            <h3
              class="crud-modal-wrapper__title"
            >
              Create a new product!
            </h3>
            <div>
              <form
                class="crud-modal-form"
              >
                
                <div
                  class="crud-modal-form__field-container"
                >
                  <label
                    class="crud-modal-form__label"
                    for="name"
                  >
                    Name
                  </label>
                  <input
                    name="name"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="crud-modal-form__field-container"
                >
                  <label
                    class="crud-modal-form__label"
                    for="type"
                  >
                    Type
                  </label>
                  <input
                    name="type"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="crud-modal-form__field-container"
                >
                  <label
                    class="crud-modal-form__label"
                    for="price"
                  >
                    Price
                  </label>
                  <input
                    name="price"
                    type="number"
                    value=""
                  />
                </div>
                <button
                  class="crud-button crud-button--positive"
                  type="submit"
                >
                  Create
                </button>
              </form>
            </div>
          </div>
        </div>
      </div>
      <div
        class="crud-table__caption"
      >
        Products
      </div>
      <table
        class="crud-table"
      >
        <thead
          class="crud-table__header"
        >
          <tr
            class="crud-table__row crud-table__row--fields"
          >
            <th
              class="crud-table__header-cell"
            >
              Id
               
              <span>
                ▼
              </span>
            </th>
            <th
              class="crud-table__header-cell"
            >
              Name
               
            </th>
            <th
              class="crud-table__header-cell"
            >
              Type
               
            </th>
            <th
              class="crud-table__header-cell"
            >
              Price
               
            </th>
            
          </tr>
          <tr
            class="crud-table__row crud-table__row--sorter"
          >
            <th
              class="crud-table__header-cell"
              colspan="4"
            >
              Sort by: 
              <br />
              <select
                class="crud-table__select"
              >
                <option
                  value=""
                >
                  Select field
                </option>
                <option
                  value="id"
                >
                  Id
                </option>
                <option
                  value="name"
                >
                  Name
                </option>
                <option
                  value="type"
                >
                  Type
                </option>
                <option
                  value="price"
                >
                  Price
                </option>
              </select>
               
              <select
                class="crud-table__select"
              >
                <option
                  value=""
                >
                  Select direction
                </option>
                <option
                  value="descending"
                >
                  descending
                </option>
                <option
                  value="ascending"
                >
                  ascending
                </option>
              </select>
            </th>
          </tr>
        </thead>
        <tbody
          class="crud-table__body"
        />
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
